fix(core): enable ansi colors console mode on windows

ostridm opened the pull request:
closes #131

ostridm committed:
fix(core): enable ANSI colors console mode on windows

closes #131

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleFormatterTests.cs:
```diff
@@ -39,7 +46,8 @@ public void Write_GivenLogLevel_LogColorSequence(LogLevel logLevel, string foreg
     });
 
-    using var sut = new ColoredConsoleFormatter(optionsMonitorMock, systemTimeProviderMock);
+
```
```suggestion
```

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleFormatterTests.cs:
```diff
@@ -39,7 +46,8 @@ public void Write_GivenLogLevel_LogColorSequence(LogLevel logLevel, string foreg
 
-    using var sut = new ColoredConsoleFormatter(optionsMonitorMock, systemTimeProviderMock);
+
+    using var sut = new ColoredConsoleFormatter(optionsMonitorMock, systemTimeProviderMock, new DefaultAnsiCodeColorizer(true));
```
Use mock instead of creating a real instance of `DefaultAnsiCodeColorizer`

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleFormatterTests.cs:
```diff
@@ -48,4 +56,65 @@ public void Write_GivenLogLevel_LogColorSequence(LogLevel logLevel, string foreg
+
+
+    using var sut = new ColoredConsoleFormatter(optionsMonitorMock, systemTimeProviderMock,
+                            new DefaultAnsiCodeColorizer(true));
```
ditto

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleFormatterTests.cs:
```diff
@@ -48,4 +56,65 @@ public void Write_GivenLogLevel_LogColorSequence(LogLevel logLevel, string foreg
+      IncludeScopes = false
+    });
+
+
```
```suggestion
```

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleFormatterTests.cs:
```diff
@@ -48,4 +56,65 @@ public void Write_GivenLogLevel_LogColorSequence(LogLevel logLevel, string foreg
+      IncludeScopes = false
+    });
+
+
```
```suggestion
```

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleFormatterTests.cs:
```diff
@@ -48,4 +56,65 @@ public void Write_GivenLogLevel_LogColorSequence(LogLevel logLevel, string foreg
+
+
+    using var sut = new ColoredConsoleFormatter(optionsMonitorMock, systemTimeProviderMock,
+      new DefaultAnsiCodeColorizer(false));
```
ditto

derevnjuk commented on test/SecTester.Core.Tests/Logger/AnsiCodeColorTests.cs:
```diff
@@ -0,0 +1,181 @@
+    },
+    new object[]
+    {
+      AnsiCodeColor.White, null as AnsiCodeColor
```
```suggestion
      AnsiCodeColor.White, null!
```

derevnjuk commented on src/SecTester.Repeater/Repeater.cs:
```diff
@@ -129,11 +135,8 @@ private void EnsureRegistrationStatus(RegisterRepeaterPayload result)
-          result.Version
-        );
+        _logger.LogWarning("{Prefix}: A new Repeater version ({Version}) is available, please update SecTester",
+          _ansiCodeColorizer.Colorize(AnsiCodeColor.Yellow, "(!) IMPORTANT"), result.Version);
```
Please colorize "(!) CRITICAL" below as well (e.g. red)

derevnjuk commented on src/SecTester.Core/Utils/ConsoleUtils.cs:
```diff
@@ -0,0 +1,61 @@
+
+  private static bool EnableWindowsAnsiColors(int consoleHandle)
+  {
+    var handle = GetStdHandle(consoleHandle);
+
+    if (handle.IsInvalid || !GetConsoleMode(handle, out var outConsoleMode))
+    {
+      return false;
+    }
+
+    return AnsiColorRequiredMode == (outConsoleMode & AnsiColorRequiredMode) ||
+           SetConsoleMode(handle, outConsoleMode | AnsiColorRequiredMode);
```
I would suggest refactoring it a bit
```csharp
private static bool EnableWindowsAnsiColors(int consoleHandle)
{
    var handle = GetStdHandle(consoleHandle);

    if (handle.IsInvalid || !GetConsoleMode(handle, out var outConsoleMode))
    {
        return false;
    }

    if ((outConsoleMode & AnsiColorRequiredMode) == AnsiColorRequiredMode)
    {
        return true;
    }

    return SetConsoleMode(handle, outConsoleMode | AnsiColorRequiredMode);
}
```

ostridm committed:
fix(core): enable ANSI colors console mode on windows

adjust according to code review:
https://github.com/NeuraLegion/sectester-net/pull/132#discussion_r1048405313
https://github.com/NeuraLegion/sectester-net/pull/132#discussion_r1048405645
https://github.com/NeuraLegion/sectester-net/pull/132#discussion_r1048405816
https://github.com/NeuraLegion/sectester-net/pull/132#discussion_r1048405876
https://github.com/NeuraLegion/sectester-net/pull/132#discussion_r1048405955
https://github.com/NeuraLegion/sectester-net/pull/132#discussion_r1048406017
https://github.com/NeuraLegion/sectester-net/pull/132#discussion_r1048406408
https://github.com/NeuraLegion/sectester-net/pull/132#discussion_r1048407672

closes #131

derevnjuk commented on test/SecTester.Core.Tests/Logger/ColoredConsoleFormatterTests.cs:
```diff
@@ -39,13 +80,19 @@ public void Write_GivenLogLevel_LogColorSequence(LogLevel logLevel, string foreg
+      .Returns(x => x.ArgAt<string>(1));
+
+    using var sut = new ColoredConsoleFormatter(optionsMonitorMock, systemTimeProviderMock,
+      new DefaultAnsiCodeColorizer(false));
```
It is still there

ostridm committed:
fix(core): enable ANSI colors console mode on windows

adjust according to code review:
https://github.com/NeuraLegion/sectester-net/pull/132#discussion_r1048451418

closes #131
